### ISSUE TITLE
Py native module import vs. nuitka-compiled module import discrepancy

### ISSUE THREAD

**Issue by milos-korenciak** [Nuitka/Nuitka#1026]
There is discrepancy when importing from a) native Py module vs b) nuitka-compiled module.

Simply run the run.sh of bundled Minimum Working Example (intended for e.g. docker ubuntu). [mwe_import_incompatibility.zip](https://github.com/Nuitka/Nuitka/files/6139600/mwe_import_incompatibility.zip)

Tested on :

- Linux (gcc) + Windows10 (Visual Studio C Compiler). The same results.
- python3.8.3 from conda
- Nuitka==0.6.12.3 (from pip)

STRONG NOTE:
First time load of even successfully loaded modules from bundled MWE would fail. The MWE therefore tries several times to load the nuitka-compiled module in try-except blocks. It DEPENDS on the order - the first import always fails.

IMHO - the importing machinery of nuitka-compiled module simply needs command to delete the partially imported module to mimic CPython behaviour.

BTW - in ipython you can see the behaviour another way (the debris of partially imported module):

```python
In [1]: import example
Exception                                 Traceback (most recent call last)
<ipython-input-...> in <module>
----> 1 import example
/tmp/mwe_import_incompatibility/example/__init__.py in <module example>()
/tmp/mwe_import_incompatibility/example/__init__.py in a()
Exception: AHOJ
In [2]: import example
In [3]: example.e
AttributeError                            Traceback (most recent call last)
<ipython-input-...> in <module>
----> 1 example.e
AttributeError: partially initialized module 'example' has no attribute 'e' (most likely due to a circular import)
```
Any help with this will be appreciated. :wink:
:thumbsup: 
Anyway - Nuitka is fine. :slightly_smiling_face:

**Comment by kayhayen** [comment]
This is pretty good as far examples go, I shall have a look once the current release process is finished, if it's small enough, that's kind of hotfix material still. Your theory is probably right. I will want to have a test that demonstrates that with CPython, and should it hold up that way, it would be easy to implement.

Something in Nuitka considers module level code as non-reentrant though, this may have other implications that I still need to think through, but on a practical level that's probably not relevant yet.

**Comment by kayhayen** [comment]
Your analysis is very nearly spot on. However, when I added `sys.modules` to the diff, btw very nice trick with the set, `example` package doesn't show up, so it does get removed. However, then when loading `example.e` it seems Nuitka loader and the import, do not re-execute the module code first indeed.

Nuitka has a flag indicating this:

```
#if defined(_NUITKA_EXE) || PYTHON_VERSION >= 0x300
    static bool _init_done = false;

    // Modules might be imported repeatedly, which is to be ignored.
    if (_init_done) {
#ifdef _NUITKA_TRACE
        PRINT_STRING("%(module_name)s: Skipping module init, already done.\n");
#endif

        return module_%(module_identifier)s;
    } else {
        _init_done = true;
    }
#endif
```

Repeated imports do not cause a new execution therefore with Python3. I did not use module mode btw, it seems that Python2 module mode, would allow re-execution, not sure if I have a test of that, but your example will be added, also as a package test. Of course, it will then be nice, if the exception was something that could be switched off for a final round.

I guess, the real trigger ought to be `sys.modules` presence, and that will be done for many cases, and what triggered this code to exist, is that some code paths hard optimize module import, in such a fashion that they jump into the C code of the compiled module without considering `sys.modules`. I think even plugin generated pre-load and post-load code for standalone modules would then run twice.

So definitely something fishy, but right now `IMPORT_EMBEDDED_MODULE` is not really used for hard imports of compiled modules in either module mode or standalone mode, althought that's clearly an omission.

Right now for 3.5 or higher `_path_unfreezer_exec_module` behaves the same though, calling the module after it was created with `_path_unfreezer_create_module` it seems. At the time it's not in sys.modules, and adding the check there would probably do, and adding that in `_EXECUTE_EMBEDDED_MODULE` might cover all code paths.

**Comment by kayhayen** [comment]
This is part of the current pre-release. Thanks for the excellent reproducer, this helped a lot to pin point the issue.

**Comment by kayhayen** [comment]
For Python3.3-Python3.5 the code on develop was not perfect yet (they are like Python2 still in their importing mechanic) and therefore develop has been rebased, or will be, for the next pre-release.

**Comment by kayhayen** [comment]
Part of the 0.6.14 release.

**Comment by milos-korenciak** [comment]
The original test failed for:
Nuitka 0.6.14.6 + py3.8.3 Lin
Nuitka 0.6.14.6 + py3.8.5 Win
I added the .bat test runner in 
[mwe_import_incompatibility.zip](https://github.com/Nuitka/Nuitka/files/6420210/mwe_import_incompatibility.zip) (but else it is the same).
Should I use any nuitka CLI switch to see the same results for native and compiled module (in the exmple)?

**Comment by kayhayen** [comment]
Ok, sorry to hear that. I isolated an issue out of it, but maybe it wasn't the only one.

**Comment by milos-korenciak** [comment]
I am in intensive debugging loop, too. :-) For me, this is just an another branch of thorough testing.
Main target is odd discrepancy between Win and Lin compilation to .so / .pyd of the very same code. :roll_eyes:
IT is a (kind of) Kinder Surprise: when you do not expect it, it will surprise you. :laughing:

**Comment by kayhayen** [comment]
This has slipped by my attention unfortunately, not sure if I can make it in time. I assume the original test case is still what I should use?

**Comment by kayhayen** [comment]
@milos-korenciak can you confirm or deny my last question?

**Comment by milos-korenciak** [comment]
Excuse me,
I have tested actual nuitka 0.6.17.6 on Py3.9.5. Still it does **not** work correctly.
Output I get:
$ bash run.sh 
run default python3.(8) test of imports
set()
set()
set()
set()
set()
set()
run nuitka-compiled test of imports - see the discrepancy
set()
set()
set()
{'a'}
{'a'}
{'a', 'Class3'}

Output I expect:
$ bash run.sh 
run default python3.(8) test of imports
set()
set()
set()
set()
set()
set()
run nuitka-compiled test of imports - see the discrepancy
set()
set()
set()
set()
set()
set()

**Comment by kayhayen** [comment]
Thanks for the feedback, I will retry your original case then. Unfortunately I isolated what I fixed only, and have to restart now.

**Comment by kayhayen** [comment]
Ok, I managed to reproduce it now. It's unfortunate that extension modules loose out against package folders it seems, so the renaming is needed. I think for files it's not.

**Comment by kayhayen** [comment]
So, your example inspired to me to add a warning about `--follow-stdlib` in module mode, and also a check that `--run` actually executes the proper code, this is naturally very important for testing.

**Comment by kayhayen** [comment]
Actually `--follow-stdlib` can make some sense, I am removing the warning, but it's definitely only for hiding purposes, like https://nuitka.net/pages/commercial.html#protection-against-reverse-enginneering

**Comment by kayhayen** [comment]
I  changed the test into a real package test for Nuitka, making `run.py` the compiled module, that includes a package that raise on import. Interestingly I didn't see the problem anymore, once I did that. I am going back to the original example to verify, but this is very curious.

**Comment by kayhayen** [comment]
BTW: I think your `run.sh` is having a glaring error, you mix `python3 run.py` and `python -m nuitka` in there, the whole point of telling people to use the `-m` is to make sure they use the same interpreter.

However, it seems that in fact, and I have to verify that, running the imports in CPython bytecode run.py behaves different that in a run as an extension module.

**Comment by kayhayen** [comment]
I am finding that Nuitka also doesn't know that imports on at least Python3 should prefer directories over extension modules, and crashes when the `example.so` exists near the `example`, when in program mode, I compile `run.py`. This is very easy stuff to get confused about.

**Comment by kayhayen** [comment]
```
uitka-Recursion:INFO: Should decide '--prefer-source-code' vs. '--no-prefer-source-code', using existing 'example' extension module by default. 

Candidates were:

ImportScanFinding(found_in='/.../mwe_import_incompatibility', priority=1, full_path='/.../mwe_import_incompatibility/example.cpython-37m-x86_64-linux-gnu.so', search_order=0) <-> 
ImportScanFinding(found_in='/.../mwe_import_incompatibility', priority=2, full_path='/.../mwe_import_incompatibility/example', search_order=0)
```

**Comment by kayhayen** [comment]
Directories with init files are to be highest priority, but they can be overtaken by python files and extension modules matching too, so that's another incompatibility fixed, as an accelerated binary, it doesn't show the issue anymore, and cannot crash from those extension modules.

**Comment by kayhayen** [comment]
I think I got much closer to explaining it, this is the traces with `--python-flag=-v` in module mode:

```
Setup nuitka compiled module/bytecode/shlib importer
...
import token # considering responsibility (find_spec)
import token # denied responsibility
Occurred AHOJ
set()
import example.e # considering responsibility (find_spec)
import example.e # claimed responsibility (compiled)
import example.e # created module
import example.e # execute module
Loaded example.e
{'Class4'}
{'e', 'Class4'}
{'e', 'a', 'Class4'}
```

Some standard imports are being done, and then the main code of the top level package is run, and that code doesn't have the thing that would remove it after error, and Python doesn't do it for the extension module it seems, so that's the point there, for modules Nuitka loads after that, it will be done, but not in case of an error.

**Comment by kayhayen** [comment]
I just pushed a fix to factory, it seems Python2 and Python3 were equally affected. I need to update my test to not do the top level runner as compiled code, not sure how to do that though, and if it's worth the effort.

**Comment by kayhayen** [comment]
https://nuitka.net/doc/factory.html

For avoiding the renames, I kept the example module in a `src` folder and added it to PYTHONPATH, or the output directory, that's then cleaner to see. I fixed a bug related to priorities, but it's messy without the separation.

**Comment by kayhayen** [comment]
Some of the correction parts I was talking about are in the hotfix, e.g. proper priorities when deciding between multiple options, but the main soliution is on the develop branch now and will be part of the next release.

**Comment by kayhayen** [comment]
Accidental close, sorry

**Comment by kayhayen** [comment]
This is part of the 0.6.18 release.
